### PR TITLE
Refactor metadata handling

### PR DESCRIPTION
**Pull request by niallcreech**
Rewriting code to move existing code into distinct classes and improve the general code base.

* Refactor existing code into logical classes
* Separate out useful functions into libraries
* Improve documentation
* Expand unit testing coverage

### CODE REVIEW

**Comment by ashb** [comment]
- [ ]  Previously I could run the effect of `salt-shaker install-pinned-versions` now without needing a github token as I was just installing the specific versions. Can we have this behaviour back please?

  (i.e. install-pinned-versions should never need to hit the GH api)

- [ ] Can you add the right thing to the dependencies to silence this warning?

   ```
   /Users/aberlin/.virtualenvs/salt-shaker-new/lib/python2.7/site-packages/requests/packages/urllib3/util/ssl_.py:90: InsecurePlatformWarning: \
A true SSLContext object is not available. \
This prevents urllib3 from configuring SSL '
appropriately and may cause certain SSL connections to fail. For more information, see \
https://urllib3.readthedocs.org/en/latest/security.html#insecureplatformwarning.
  InsecurePlatformWarning
   ```

  (line wraps were added by me to make it easier to read on github)

- [ ] Running `salt-shaker install` I also see some warnings about `WARNING:salt-shaker:validate_github_access: URL not found`

  During normal operation I shouldn't see these unless it is actually a problem else it is just noise and means nothing to me as a user of the tool.

- [ ] Log output is too noisey and confusing by default

   ```
   WARNING:salt-shaker:validate_github_access: URL not found
   WARNING:salt-shaker:ShakerMetadata::_fetch_remote_requirements: No requirements found for ministryofjustice/bootstrap-formula
   WARNING:salt-shaker:validate_github_access: URL not found
   WARNING:salt-shaker:validate_github_access: URL not found
   WARNING:salt-shaker:ShakerMetadata::_fetch_remote_requirements: No requirements found for ministryofjustice/apparmor-formula
   WARNING:salt-shaker:validate_github_access: URL not found
   WARNING:salt-shaker:validate_github_access: URL not found
   ```

- [ ] What does the "No requirements found" mean? Not sure it's useful to say. (as in not sure either way on this one)

- [ ] I tried creating this line in my metadata.yml file:

   ```
     - ministryofjustice/ntp-formula < 1.4
   ```

   And I got a weird unhelpful error: 

   ```
   Traceback (most recent call last):
     File "/Users/aberlin/.virtualenvs/salt-shaker-new/bin/salt-shaker", line 6, in <module>
       ShakerCommandLine().run(sys.argv[1:])
     File "/Users/aberlin/.virtualenvs/salt-shaker-new/lib/python2.7/site-packages/shaker/command_line.py", line 46, in run
       return args.pop('func')(**args)
     File "/Users/aberlin/.virtualenvs/salt-shaker-new/lib/python2.7/site-packages/shaker/command_line.py", line 58, in shake
       salt_shaker.shaker(**kwargs)
     File "/Users/aberlin/.virtualenvs/salt-shaker-new/lib/python2.7/site-packages/shaker/salt_shaker.py", line 160, in shaker
       shaker_instance.update_requirements()
     File "/Users/aberlin/.virtualenvs/salt-shaker-new/lib/python2.7/site-packages/shaker/salt_shaker.py", line 85, in update_requirements
       self._shaker_metadata.update_dependencies(ignore_local_requirements=True)
     File "/Users/aberlin/.virtualenvs/salt-shaker-new/lib/python2.7/site-packages/shaker/shaker_metadata.py", line 152, in update_dependencies
       ignore_dependency_requirements)
     File "/Users/aberlin/.virtualenvs/salt-shaker-new/lib/python2.7/site-packages/shaker/shaker_metadata.py", line 450, in _fetch_dependencies
       constraint=constraint)
     File "/Users/aberlin/.virtualenvs/salt-shaker-new/lib/python2.7/site-packages/shaker/shaker_metadata.py", line 566, in _fetch_remote_requirements
       constraint)
     File "/Users/aberlin/.virtualenvs/salt-shaker-new/lib/python2.7/site-packages/shaker/shaker_metadata.py", line 625, in _fetch_remote_file
       raise GithubRepositoryConnectionException(msg)
   shaker.libs.errors.GithubRepositoryConnectionException: ShakerMetadata::_fetch_remote_file: ministryofjustice/ntp-formula < 1.4:: No target object found, check it exists and you have the environment variable GITHUB_TOKEN set for authenticated access to private repositories
   ```

**Comment by bennythejudge** [comment]
I will be looking at the following points in new branch
- Running salt-shaker install I also see some warnings about WARNING:salt-shaker:validate_github_access: URL not found
-  Log output is too noisey and confusing by default
- I tried creating this line in my metadata.yml file (...) (invalid constraint)

**Comment by niallcreech** [comment]
salt-shaker install-pinned-versions uses the github api to take the tag in the formula-requirements and resolve it to a sha, on the basis that even though the tag hasn't changed the sha it points to may well have. Is there a way round this?

**Comment by ashb** [comment]
@niallcreech We can and should assume that tags are immutable. Moving tags is very bad form. so we can just check out a repo and make sure its at what ever that tag name is - we don't need to resolve it direclty to a sha, git handles that for us.

**Comment by niallcreech** [comment]
Can do, I was working on the principle that tags might move, I'll look at the more straightfoward (fingers-crossed) case

File: README.md [hunk 1]
```diff
@@ -7,8 +9,69 @@ Note: Install libgit2 with libssh2 support before installing this package.
...
+    
+Salt shakers requires an initial config file containing the metadata for the local formula. eg,
+
+	formula: my_organisation/local-formula
+
+	dependencies:
+		- some_organisation/test1-formula
+		- another_organisation/testa-formula>=v1.0.0
+		- another_organisation/testb-formula<=v4.0.0
+		- another_organisation/testc-formula==v2.0.0
```
Comment (ashb) [inline README.md]: These should probably be in a code block like this:

	```
	formula: my_organisation/local-formula

	dependencies:
		- some_organisation/test1-formula
		- another_organisation/testa-formula>=v1.0.0
		- another_organisation/testb-formula<=v4.0.0
		- another_organisation/testc-formula==v2.0.0
	```

Comment (ashb) [inline README.md]: And this needs updating - I think `formula` isn't a key that is used anymore?

File: README.md [hunk 2]
```diff
@@ -7,8 +9,69 @@ Note: Install libgit2 with libssh2 support before installing this package.
...
+		
+To generate and download a list of formula requirements, simply run
+
+	salt-shaker update
+	
+This will also save a list of the requirements and their versions, by default in 'formula-requirements.txt'
+
+If this file exists, you can run 
+
+	salt-shaker refresh
```
Comment (ashb) [inline README.md]: These commands don't exist - please update them to reflect the name of the commands as they now are

File: README.md [hunk 3]
```diff
@@ -7,8 +9,69 @@ Note: Install libgit2 with libssh2 support before installing this package.
...
+
+Salt shakers requires an initial config file containing the metadata for the local formula. eg,
+
+	formula: my_organisation/local-formula
+
+	dependencies:
+		- some_organisation/test1-formula
+		- another_organisation/testa-formula>=v1.0.0
+		- another_organisation/testb-formula<=v4.0.0
+		- another_organisation/testc-formula==v2.0.0
```
Comment (ashb) [inline README.md]: Same here - update and make a code block

File: README.md (hunk 1 above)
Comment (niallcreech) [inline README.md]: It was 'entry' that was dropped. 'formula' is used for the root name

File: README.md (hunk 2 above)
Comment (niallcreech) [inline README.md]: Updated the docs on the make_api_back_compatible branch

File: README.md (hunk 3 above)
Comment (niallcreech) [inline README.md]: Updated the docs on the make_api_back_compatible branch

File: setup.py [hunk 4]
```diff
@@ -13,10 +14,11 @@
         'requests',
```
Comment (ashb) [inline setup.py]: This needs changing to `requests[security]` as this is what is used when we do `pip install salt-shaker` – it won't look at requirements.txt in that case.

We should probably remove requirements.txt entirely.

**Comment by niallcreech** [comment]
Update setup.py

**Comment by niallcreech** [comment]
rebased to master

**Comment by niallcreech** [comment]
The links to the formulas in _root don't appear to be right. ie,
docker-formula, not docker
will sort out and resubmit

**Comment by niallcreech** [comment]
Requires https://github.com/ministryofjustice/salt-shaker/pull/42

**Pull request merged**